src/lib.rs

```diff
@@ -191,83 +230,133 @@ pub mod compiler_pipeline {
                               file: &str,
                               expr_str: &str,
                               expected_type: Option<&ArcType>)
-                              -> Result<TypecheckValue>;
+                              -> Result<TypecheckValue<Self::Expr>>;
     }
+
     impl<T> Typecheckable for T
         where T: MacroExpandable,
     {
+        type Expr = T::Expr;
+
         fn typecheck_expected(self,
                               compiler: &mut Compiler,
                               thread: &Thread,
                               file: &str,
                               expr_str: &str,
                               expected_type: Option<&ArcType>)
-                              -> Result<TypecheckValue>
-            where Self: Sized,
-        {
+                              -> Result<TypecheckValue<Self::Expr>> {
             self.expand_macro(compiler, thread, file)
                 .and_then(|expr| {
                     expr.typecheck_expected(compiler, thread, file, expr_str, expected_type)
                 })
         }
     }
-    impl Typecheckable for MacroValue {
+
+    impl<E> Typecheckable for MacroValue<E>
+        where E: BorrowMut<SpannedExpr<Symbol>>,
+    {
+        type Expr = E;
+
         fn typecheck_expected(mut self,
                               compiler: &mut Compiler,
                               thread: &Thread,
                               file: &str,
                               expr_str: &str,
                               expected_type: Option<&ArcType>)
-                              -> Result<TypecheckValue>
-            where Self: Sized,
-        {
-            compiler.typecheck_expr_expected(thread, file, expr_str, &mut self.0, expected_type)
-                .map(move |typ| TypecheckValue(self.0, typ))
+                              -> Result<TypecheckValue<Self::Expr>> {
+            use check::typecheck::Typecheck;
+
+            let env = thread.get_env();
+            let mut tc = Typecheck::new(file.into(), &mut compiler.symbols, &*env);
+
+            let typ = try!(tc.typecheck_expr_expected(self.expr.borrow_mut(), expected_type)
+                .map_err(|err| InFile::new(file, expr_str, err)));
+
+            Ok(TypecheckValue {
+                expr: self.expr,
+                typ: typ,
+            })
         }
     }
 
-    pub struct CompileValue(pub SpannedExpr<Symbol>, pub ArcType, pub CompiledFunction);
+    /// Result of successful compilation
+    pub struct CompileValue<E> {
+        pub expr: E,
+        pub typ: ArcType,
+        pub function: CompiledFunction,
+    }
 
     pub trait Compileable<Extra> {
+        type Expr;
```

Comment:
Inconsistent new-line spacing here. I would add one between the type and the following function.